bridge/prometheus/producer.go

```diff
@@ -0,0 +1,253 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus // import "go.opentelemetry.io/otel/bridge/prometheus"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+const (
+	scopeName    = "go.opentelemetry.io/otel/bridge/prometheus"
+	traceIDLabel = "trace_id"
+	spanIDLabel  = "span_id"
+)
+
+var (
+	errUnsupportedType = errors.New("unsupported metric type")
+	processStartTime   = time.Now()
+)
+
+type producer struct {
+	gatherers []prometheus.Gatherer
+}
+
+// NewMetricProducer returns a metric.Producer that fetches metrics from
+// Prometheus. This can be used to allow Prometheus instrumentation to be
+// added to an OpenTelemetry export pipeline.
+func NewMetricProducer(opts ...Option) metric.Producer {
+	cfg := newConfig(opts...)
+	return &producer{
+		gatherers: cfg.gatherers,
+	}
+}
+
+func (p *producer) Produce(context.Context) ([]metricdata.ScopeMetrics, error) {
+	now := time.Now()
+	var errs multierr
+	otelMetrics := make([]metricdata.Metrics, 0)
+	for _, gatherer := range p.gatherers {
+		promMetrics, err := gatherer.Gather()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		m, err := convertPrometheusMetricsInto(promMetrics, now)
+		otelMetrics = append(otelMetrics, m...)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if errs.errOrNil() != nil {
+		otel.Handle(errs.errOrNil())
+	}
+	if len(otelMetrics) == 0 {
+		return nil, nil
+	}
+	return []metricdata.ScopeMetrics{{
+		Scope: instrumentation.Scope{
+			Name: scopeName,
+		},
+		Metrics: otelMetrics,
+	}}, nil
+}
+
+func convertPrometheusMetricsInto(promMetrics []*dto.MetricFamily, now time.Time) ([]metricdata.Metrics, error) {
+	var errs multierr
+	otelMetrics := make([]metricdata.Metrics, 0)
+	for _, pm := range promMetrics {
+		newMetric := metricdata.Metrics{
+			Name:        pm.GetName(),
```

Comment:
Is this necessary to convert prometheus metric name to otlp metrics name?
For example convert `http_request_duration_count` to `http.request.duration.count`?

Reply:
Unfortunately, it isn't feasible to do that, as otel can have underscores in names as well (e.g. `my.very_special.metric`)